bee-network/bee-autopeering/src/packet.rs

```diff
@@ -0,0 +1,127 @@
+
+use base64 as bs64;
+use crypto::{
+    hashes::sha::SHA256_LEN,
+    signatures::ed25519::{PublicKey, Signature},
+};
+use num_derive::FromPrimitive;
+use prost::{bytes::BytesMut, DecodeError, EncodeError, Message};
+
+use std::{fmt, net::SocketAddr, ops::Range};
+
+// From `hive.go` docs:
+// * specifies the maximum allowed size of packets;
+// * packets larger than this will be cut and thus treated as invalid;
+pub(crate) const MAX_PACKET_SIZE: usize = 1280;
+
+pub(crate) const DISCOVERY_MSG_TYPE_MIN: u32 = 10;
+pub(crate) const DISCOVERY_MSG_TYPE_RANGE: Range<u32> = DISCOVERY_MSG_TYPE_MIN..(DISCOVERY_MSG_TYPE_MIN + 4);
+pub(crate) const PEERING_MSG_TYPE_MIN: u32 = 20;
+pub(crate) const PEERING_MSG_TYPE_RANGE: Range<u32> = PEERING_MSG_TYPE_MIN..(PEERING_MSG_TYPE_MIN + 3);
+
+pub(crate) fn msg_hash(msg_type: MessageType, msg_data: &[u8]) -> [u8; SHA256_LEN] {
+    let mut bytes = vec![0u8; msg_data.len() + 1];
+    let msg_type = msg_type as u32;
+    if msg_type > 0xFF {
+        panic!("invalid message type");
+    }
+    bytes[0] = msg_type as u8;
+    bytes[1..].copy_from_slice(msg_data);
+
+    hash::sha256(&bytes)
+}
+
+/// Represents an IOTA packet.
+pub(crate) struct Packet(proto::Packet);
+
+impl Packet {
+    /// Creates a new packet.
+    pub(crate) fn new(msg_type: MessageType, msg_bytes: &[u8], public_key: &PublicKey, signature: Signature) -> Self {
+        Self(proto::Packet {
+            r#type: msg_type as u32,
+            data: msg_bytes.to_vec(),
+            public_key: public_key.to_bytes().to_vec(),
+            signature: signature.to_bytes().to_vec(),
+        })
+    }
+
+    /// Returns the message contained in this packet.
+    pub(crate) fn message(&self) -> &[u8] {
+        &self.0.data
+    }
+
+    /// Returns the public key belonging to the issuer of this packet.
+    pub(crate) fn public_key(&self) -> PublicKey {
+        PublicKey::try_from_bytes(self.0.public_key.clone().try_into().expect("error public key length"))
+            .expect("error restoring public key from bytes")
+    }
+
+    /// Returns the signature belonging to the issuer of this packet.
+    pub(crate) fn signature(&self) -> Signature {
+        Signature::from_bytes(self.0.signature.clone().try_into().expect("error signature length"))
+    }
+
+    /// Restores a packet from its protobuf representation.
+    pub(crate) fn from_protobuf(bytes: &[u8]) -> Result<Self, DecodeError> {
+        Ok(Self(proto::Packet::decode(bytes)?))
+    }
+
+    /// Returns the protobuf representation of this packet
+    pub(crate) fn to_protobuf(&self) -> Result<BytesMut, EncodeError> {
+        let mut buf = BytesMut::with_capacity(self.0.encoded_len());
+        self.0.encode(&mut buf)?;
+
+        Ok(buf)
+    }
+
+    /// Turns the packet into its contained message (if any) and discards the rest of the metadata.
+    pub(crate) fn into_message(self) -> Vec<u8> {
+        self.0.data
+    }
+}
+
+impl fmt::Debug for Packet {
```

Comment:
You can ignore this, just raising that this seems expensive so it probably shouldn't be in a hot path

Reply:
right it shouldn't be in a hot path, and hopefully it isn't. A quick search didn't yield it being in a trace log or whatever. This is really just there because it should be there in case we need to debug the socket I/O.